torch/distributed/fsdp/_init_utils.py

```diff
@@ -75,13 +88,140 @@
 @no_type_check
 def _init_process_group_state(
     state: _FSDPState,
-    process_group: Optional[dist.ProcessGroup],
+    process_group: ProcessGroupType,
+    sharding_strategy: ShardingStrategy,
+    policy: Optional[_FSDPPolicy],
 ) -> _FSDPState:
-    state.process_group = process_group or _get_default_group()
+    if sharding_strategy in HYBRID_SHARDING_STRATEGIES:
+        if process_group is None and policy is None:
+            # Raise an error here, since this is manual wrapping with no process group
+            # passed in, there is no way to ensure all wrapped FSDP units use the same
+            # process groups.
+            raise ValueError(
+                f"Manual wrapping with {sharding_strategy} requires explicit specification of process group."
+            )
+        else:
+            state = _init_process_group_state_for_hybrid_shard(state, process_group)
+            assert state.process_group is not None, "Expected to populate state.process_group for hybrid shard"
+            assert state._inter_node_pg is not None, "Expected to populate state._inter_node_pg for hybrid shard"
+            assert state._inter_node_state is not None, "Expected to populate state._inter_node_state for hybrid shad."
+    else:
+        if process_group is not None and not isinstance(process_group, dist.ProcessGroup):
+            raise ValueError(
+                f"process_group should be None or dist.ProcessGroup, but got {type(process_group)}"
+            )
+        if process_group is None:
+            state.process_group = _get_default_group()
+        else:
+            state.process_group = process_group
+
     state.rank = state.process_group.rank()
     state.world_size = state.process_group.size()
+
     return state
 
+@no_type_check
+def _init_process_group_state_for_hybrid_shard(state, process_group) -> _FSDPState:
+    if process_group is None:
+        default_group = _get_default_group()
+        intra_node_group, inter_node_group = _init_intra_and_inter_node_groups(default_group)
+        # we shard across intra-node
+        state.process_group = intra_node_group
+        # save _inter_node_pg to allreduce across.
+        state._inter_node_pg = inter_node_group
+    else:
+        # Check type and assign state.process_group and state._inter_node_pg.
+        valid = _validate_hybrid_shard_pg_type(process_group)
+        if valid:
```

Comment:
Following other comment below:
```suggestion
        if _is_valid_hybrid_shard_pg_type(process_group):
```

(I like this kind of naming because then it reads like English.)

Reply:
[Easy] I recommend changing before landing